### PR TITLE
Implement OSPFv3 protocol

### LINKED ISSUE #143: Add OSPF Area 0 support
OSPFv2/v3 with Area 0 support only

### PR DESCRIPTION
**Pull request by mraerino**
*Fixes #143*

### Relevant RFCs
- [RFC 5340 - OSPF for IPv6](https://tools.ietf.org/html/rfc5340)
- [RFC 5838 - Support of Address Families in OSPFv3](https://tools.ietf.org/html/rfc5838)
- [RFC 2328 - OSPF Version 2](https://tools.ietf.org/html/rfc2328)

### Steps
- [x] Lab OSPFv3 setup with bird
- [x] Generate & implement packet decoding according to [RFC 5340](https://tools.ietf.org/html/rfc5340)
- [ ] Support legacy IP in OSPFv3 ([RFC 5838](https://tools.ietf.org/html/rfc5838))
- [ ] Implement establishment of router adjacency with neighbor
- [ ] Implement link database updates **for Area 0 only**
- [ ] Implement full area support

### Trying it out

The decoding of OSPFv3 messages can be tried on packet captures from [packetlife.net](https://packetlife.net/captures/protocol/ospf/):
```bash
go run examples/ospf/decode/main.go
```

Any early feedback is appreciated!

### CODE REVIEW

File: protocols/ospfv3/packet/common.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	buf.Write(convert.Uint16Byte(uint16(r.Flags)))
+}
+
+type LSType uint16
+
+func (t LSType) Serialize(buf *bytes.Buffer) {
+	buf.Write(convert.Uint16Byte(uint16(t)))
+}
+
+type IPv6Address struct {
```
Comment (taktv6) [inline protocols/ospfv3/packet/common.go]: Please use IP type from net package

File: protocols/ospfv3/packet/lsa.go [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+	AdvertisingRouter ID
+	SequenceNumber    uint32
+	Checksum          uint16
+	Length            uint16
+	Body              Serializable
+}
+
+const LSAHeaderLength = 20
+
+func (x *LSA) Serialize(buf *bytes.Buffer, headerOnly bool) {
```
Comment (taktv6) [inline protocols/ospfv3/packet/lsa.go]: "A flag argument is a kind of function argument that tells the function to carry out a different operation depending on its value." https://martinfowler.com/bliki/FlagArgument.html
Let's make it 2 functions.

File: protocols/ospfv3/packet/lsa.go [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+	x.AdvertisingRouter.Serialize(buf)
+	buf.Write(convert.Uint32Byte(x.SequenceNumber))
+	buf.Write(convert.Uint16Byte(x.Checksum))
+	buf.Write(convert.Uint16Byte(x.Length))
+	if !headerOnly {
+		x.Body.Serialize(buf)
+	}
+}
+
+func DeserializeLSA(buf *bytes.Buffer, headerOnly bool) (*LSA, int, error) {
```
Comment (taktv6) [inline protocols/ospfv3/packet/lsa.go]: Same here. No bool parameters please.

File: protocols/ospfv3/packet/lsa.go [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+
+func (x *LSA) ReadBody(buf *bytes.Buffer) (int, error) {
+	code := x.Type & 8191 // Bitmask excludes top three bits
+	bodyLength := x.Length - LSAHeaderLength
+	var body Serializable
+	var readBytes int
+	var err error
+
+	switch code {
+	case 1:
```
Comment (taktv6) [inline protocols/ospfv3/packet/lsa.go]: magic numbers. Let's give them names.

File: examples/ospf/decode/main.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
+package main
```
Comment (taktv6) [inline examples/ospf/decode/main.go]: Let's make this a test case?

**Comment by mraerino** [comment]
Superseded by https://github.com/bio-routing/bio-rd/pull/225

**Pull request closed without merging**